src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java

```diff
@@ -283,6 +296,81 @@ private static void deleteHadoopFile(String fileToObliterate, Configuration conf
         pathToDelete.getFileSystem(conf).delete(pathToDelete, true);
     }
 
+    private static void mergeHeaderlessBamShards(final JavaSparkContext ctx, final String outputFile, final SAMFormat samOutputFormat, final SAMFileHeader header) throws IOException {
+        // At this point, the part files (part-r-00000, part-r-00001, etc) are in a directory named outputFile.
+        // Each part file is a BAM file with no header or terminating end-of-file marker (Hadoop-BAM does not add
+        // end-of-file markers), so to merge into a single BAM we concatenate the header with the part files and a
+        // terminating end-of-file marker. Note that this has the side effect of being ever-so-slightly less efficient
+        // than writing a BAM in one go because the last block of each file isn't completely full.
+
+        final String outputParentDir = outputFile.substring(0, outputFile.lastIndexOf('/') + 1);
+        // First, check for the _SUCCESS file.
+        final String successFile = outputFile + "/_SUCCESS";
+        final Path successPath = new Path(successFile);
```

Comment:
Perhaps we should use a fully-qualified class name when using Hadoop's Path (`org.apache.hadoop.fs.Path`), to avoid potential confusion with NIO Path?